src/components/application_manager/include/application_manager/system_time/system_time_handler_impl.h

```diff
@@ -0,0 +1,151 @@
+#include "utils/system_time_handler.h"
+#include "application_manager/application_manager_impl.h"
+#include "application_manager/event_engine/event_observer.h"
+
+namespace application_manager {
+
+/**
+ * @brief The SystemTimeHandlerImpl class.
+ * Responsible for the system time notification. It keeps the list
+ * of system time listeners, sends appropriate request to the system
+ * and notifies the listeners with new time right after response has appeared.
+ */
+class SystemTimeHandlerImpl : public utils::SystemTimeHandler,
+                              public event_engine::EventObserver {
+ public:
+  /**
+   * @brief SystemTimeHandlerImpl creates the instance of the class.
+   */
+  explicit SystemTimeHandlerImpl(ApplicationManager& application_manager);
+
+  /**
+   * @brief ~SystemTimeHandlerImpl
+   */
+  ~SystemTimeHandlerImpl();
+
+ private:
+  /**
+   * @brief on_event allows to process certain events from the system.
+   * Event which are handles within this methods  are: OnSystemTimeReady
+   * in order to send system time query and GetSystemTimeResponse in order
+   * to retrieve utc time and notify all the listeners with new time value.
+   */
+  void on_event(const application_manager::event_engine::Event& event) FINAL;
+
+  /**
+   * @brief DoSystemTimeQuery sends the appropriate request to the system
+   * and subscribes the class to the response. This is asynchronous request
+   * thus it won't block until the system respond and returns immediately.
+   */
+  void DoSystemTimeQuery() FINAL;
+
+  /**
+   * @brief DoSubscribe allows to subscribe certain listener for
+   * SystemTime. This certain implementation does not maintain
+   * listeners collection instead it save the pointer to listener.
+   * @note the class is not reponsible for the pointer's lifecycle.
+   * So consider to explicitly unsibsscribe from event when listener
+   * is going to be destroyed.
+   * @param listener the listener which will be notified
+   * in case of SystemTime appeared.
+   */
+  void DoSubscribe(utils::SystemTimeListener* listener) FINAL;
+
+  /**
+   * @brief DoUnsubscribe assigns the pointer to listener to NULL.
+   * This certain implementation ignores passed parameter.
+   * @param listener pointer to the object which should be
+   * unsubscribed from events.
+   */
+  void DoUnsubscribe(utils::SystemTimeListener* listener) FINAL;
+
+  /**
+   * @brief FetchSystemTime this method allows to obtain
+   * recently received utc time.
+   * @note it is up to user to check whether the returned
+   * argument is valid
+   * @return the time value returned by system.
+   */
+  time_t FetchSystemTime() FINAL;
+
+  /**
+   * @brief SendTimeRequest sends the request to the system
+   * and subscribes for response.
+   */
+  void SendTimeRequest();
+
+  /**
+   * @brief OnSystemTimeResponse allows to process GetSystemTimeResponse
+   * @param event contains response parameters aka "systemTime".
+   */
+  void OnSystemTimeResponse(
```

Comment:
@AByzhynar please specify, Please remove `On` prefix there, we already have OnEvent, rename it to `ProcessSystemTimeResponse`